agents/mini_max.py

```diff
@@ -0,0 +1,46 @@
+
+class MiniMax():
+    """
+    General implementation of MiniMax algorithm
+    
+    a singleton class with one public method: get Action
+
+    get_action: 
+      - responsible for returning th minmaxed move given a state
+      - params:
+            - generate_children: a function that takes a state and returns a list of children states
+                    - should be a list of states, where each state is a tuple of (board, player, adversary, max step)
```

Comment:
Why does max step need to be included?

Reply:
The generate_children() function will need to know how many steps to take
so the general state needs to include
(board, player, adversary, max_step) so generate children  knows what it can generate, because it will call get possible moves

Not completely necessary, if our generate_children() funciton some how accounts for this and knows about it